enrich/enricher.go

```diff
@@ -56,69 +73,155 @@ func (e *Enricher) GetEDCommon(ctx context.Context, logGroup, accountID string)
+
 	return &Common{
-		Cloud: &cloud{ResourceID: functionARN},
+		Cloud: &cloud{ResourceID: getResourceID(arnsToGetTags, forwarderARN, logGroupARN), AccountID: accountID, Region: e.region},
 		Faas: &faas{
-			Name:    functionName,
-			Version: functionVersion,
+			Name:       functionName,
+			Version:    functionVersion,
+			RequestID:  lc.AwsRequestID,
+			MemorySize: memorySize,
+			Tags:       faasTags,
 		},
-		LambdaTags: tags,
+		AwsCommon: &awsCommon{
+			LogGroup:  logGroup,
+			LogStream: logStream,
+		},
+		HostArchitecture:   hostArchitecture,
+		ProcessRuntimeName: processRuntimeName,
+		Tags:               tags,
+	}
+}
+
+func (e *Enricher) getAllTags(ctx context.Context, forwarderARN, logGroupARN string, allARNs []string, isSourceLambda bool) (map[string]string, map[string]string) {
+	e.prepareResourceTags(ctx, allARNs)
+	tags := map[string]string{}
+	if m, ok := resourceTagsCache[logGroupARN]; ok {
+		for k, v := range m {
+			tags[k] = v
+		}
+	}
+	faasTags := map[string]string{}
+	if m, ok := resourceTagsCache[forwarderARN]; ok {
+		for k, v := range m {
+			faasTags[k] = v
+		}
+	}
+
+	sourceTags := tags
+	if isSourceLambda {
+		sourceTags = faasTags
 	}
+	for _, arn := range allARNs {
+		if arn == forwarderARN || arn == logGroupARN {
+			continue
+		}
+		if m, ok := resourceTagsCache[arn]; ok {
+			for k, v := range m {
+				sourceTags[k] = v
+			}
+		}
+	}
+	return tags, faasTags
 }
 
-func (e *Enricher) getLambdaTags(ctx context.Context, functionARN string) map[string]string {
-	if tags, ok := resourceARNToTagsCache[functionARN]; ok {
-		return tags
+func (e *Enricher) prepareResourceTags(ctx context.Context, arns []string) {
+	tagsCacheKey := getTagsCacheKey(arns...)
+	if _, ok := resourceTagsCache[tagsCacheKey]; ok {
+		return
 	}
-	tagsMap, err := e.resourceCl.GetResourceTags(ctx, functionARN)
+	log.Printf("Getting resource tags for ARNs: %v", arns)
+	tagsMap, err := e.resourceCl.GetResourceTags(ctx, arns...)
 	if err != nil {
-		log.Printf("Failed to get resource tags for ARN: %s, err: %v", functionARN, err)
-	} else if len(tagsMap) == 0 {
-		log.Printf("Failed to find tags for ARN: %s", functionARN)
-	} else {
-		for r, t := range tagsMap {
-			log.Printf("Found tags: %v for ARN: %s", t, r)
-			resourceARNToTagsCache[r] = t
+		log.Printf("Failed to get resource tags for ARNs: %v, err: %v", arns, err)
+		return
+	}
+	if len(tagsMap) == 0 {
+		log.Printf("Failed to find tags for ARNs: %v", arns)
+		return
+	}
+
+	resourceTagsCache[tagsCacheKey] = map[string]string{}
+	for r, t := range tagsMap {
+		resourceTagsCache[r] = t
+		log.Printf("Found tags: %v for ARN: %s", t, r)
+		for k, v := range t {
+			resourceTagsCache[tagsCacheKey][k] = v
 		}
 	}
-	return resourceARNToTagsCache[functionARN]
 }
 
-func getFunctionName(logGroup string) (string, bool) {
-	name := strings.TrimPrefix(logGroup, lambdaLogGroupPrefix)
-	if len(name) == len(logGroup) {
-		return "", false
+func getTagsCacheKey(arns ...string) string {
```

Comment:
I think we should cache tags per arn

Reply:
Actually we cache per ARN but since only one request is sent I wanted to put a map as well for ARN combination to avoid additional requests